### DISCUSSION TITLE
GitHub Disaster Recovery - gaps in branch restore etc.

### DISCUSSION THREAD

**Discussion by HariSekhon** [community/community#16310] [Pull Requests]
I've been working on some aspects of GitHub DR for my enterprise, and while some parts like undeleting a repo, or undeleting a branch that was deleted via a pull request or the UI work quite well, there are still some gaps.

For example, if you delete a branch before a pull request is raised via something like:

```
git push origin :somebranch
```

then the branch and its commits are lost in GitHub.

If you don't have a local copy of the branch, then any commits in that branch are permanently lost. I suspect the same is true of API deletes of repositories but haven't tested this yet.

I use branch protections on major branches, but it's still possible that new feature branches could be lost with significant pieces of work through human error.

It's also possible that a simple commenting out in Terraform could delete branches including their branch protections due to [this issue](https://github.com/hashicorp/terraform/issues/17599), or potentially even entire repos.

I don't believe GitHub's DR strategy is sufficient at this time, and I even raised a premium support request which pointed me to try the 3rd party solutions like BackHub, which I did, but found it could not restore pull requests properly due to limitations with the GitHub API, and could not configure a backup schedule aside from once a day, too high an RPO for a busy repo in my opinion.

Basically, there is no good comprehensive DR solution for GitHub today.

I end up doing my own DR solution for GitHub by mirroring to GitLab, Bitbucket, Azure Devops, GCP Source Repos and AWS CodeCommit, via scripts in my DevOps Bash tools repo and my GitHub Actions workflows reusable library:

https://github.com/HariSekhon/DevOps-Bash-tools

https://github.com/HariSekhon/GitHub-Actions

Please can you provide a more comprehensive disaster recovery solution to address these points without needing 3rd parties?
